plugins/com.google.cloud.tools.eclipse.appengine.libraries.ui/src/com/google/cloud/tools/eclipse/appengine/libraries/ui/CloudLibrariesPage.java

```diff
@@ -18,106 +18,197 @@
   }
 
   @Override
   public void createControl(Composite parent) {
-    Composite composite = new Composite(parent, SWT.BORDER);
-    composite.setLayout(new GridLayout(2, true));
-    
-    boolean java7AppEngineStandardProject = false;
+    Preconditions.checkNotNull(libraryGroups, "Library groups must be set"); //$NON-NLS-1$
+    Composite composite = new Group(parent, SWT.NONE);
+
     IProjectFacetVersion facetVersion =
         AppEngineStandardFacet.getProjectFacetVersion(project.getProject());
-    if (facetVersion != null && facetVersion.getVersionString().equals("JRE7")) {
-      java7AppEngineStandardProject = true;
+    boolean java7AppEngineStandardProject = AppEngineStandardFacet.JRE7.equals(facetVersion);
+
+    // create the library selector libraryGroups
+    for (Entry<String, String> group : libraryGroups.entrySet()) {
+      LibrarySelectorGroup librariesSelector =
+          new LibrarySelectorGroup(composite, group.getKey(), group.getValue(),
+              java7AppEngineStandardProject);
+      librariesSelectors.add(librariesSelector);
     }
-    
-    librariesSelector = new LibrarySelectorGroup(composite, group, java7AppEngineStandardProject);
-    
+    setSelectedLibraries(initialSelection);
+    composite.setLayout(new RowLayout(SWT.HORIZONTAL));
     setControl(composite);
   }
 
   @Override
   public boolean finish() {
-    return true;
+    final List<Library> libraries = getSelectedLibraries();
+    try {
+      if (isMavenProject) {
+        BuildPath.addMavenLibraries(project.getProject(), libraries, new NullProgressMonitor());
+      } else {
+        /*
+         * FIXME: BuildPath.addNativeLibrary() is too heavy-weight here. ClasspathContainerWizard,
+         * our wizard, is responsible for installing the classpath entry returned by getSelection(),
+         * which will perform the library resolution. We just need to save the selected libraries 
+         * so that they are resolved later.
+         */
+        Library masterLibrary =
+            BuildPath.collectLibraryFiles(project, libraries, new NullProgressMonitor());
+        newEntry = BuildPath.computeEntry(project, masterLibrary, new NullProgressMonitor());
+        BuildPath.saveLibraryList(project, libraries, new NullProgressMonitor());
+        if (newEntry == null) {
+          // container-editing only refreshes the content if new
+          BuildPath.runContainerResolverJob(project);
+        }
+      }
+      return true;
+    } catch (CoreException ex) {
+      StatusUtil.setErrorStatus(this, "Error updating container definition", ex); //$NON-NLS-1$
+      return false;
+    }
   }
 
-  @Override
-  public IClasspathEntry getSelection() {
-    // Since this class implements IClasspathContainerPageExtension2,
-    // Eclipse calls getNewContainers instead.
-    logger.log(Level.WARNING, "Unexpected call to getSelection()");
-    return null;
+  @VisibleForTesting
+  List<Library> getVisibleLibraries() {
```

Comment:
This is used only in testing. You can remove this and do

```java
    Predicate<Control> isObjectify = new Predicate<Control>() {
      @Override
      public boolean apply(Control control) {
        return control instanceof Button && ((Button) control).getText().equals("Objectify");
      }
    };
    assertNotNull(CompositeUtil.findControl(shellTestResource.getShell(), isObjectify));
```
or
```java
    assertNull(CompositeUtil.findControl(shellTestResource.getShell(), isObjectify));
```

Reply:
You should be able to remove `CloudLibrariesGroup.getAvailableLibraries()` too.

Reply:
That SWT grovelling code is a bit icky, don't you think?

But I can rework the tests to just look for the library group rather than objectify specifically.